--- old/LhA.c	2006-06-22 00:22:23.000000000 +0300
+++ clients/LhA.c	2006-06-18 17:51:30.000000000 +0300
@@ -92,7 +92,7 @@
   xadSTRPTR filename2 = 0, comment2 = 0;
   xadINT32 namesize = 0, dirsize = 0, commentsize = 0, groupsize = 0, usersize = 0;
   xadUINT32 userid = 0, groupid = 0;
-  xadINT32 time = 0, prot, err = 0;
+  xadINT32 time = 0, prot = -1, err = 0;
   xadUINT8 buf[128];
   xadSTRPTR buf2 = 0;
   xadUINT8 *bufptr;
@@ -128,7 +128,7 @@
       {
 //      case 0x00: /* CRC-16 of header */ break;
       case 0x01: /* Filename */
-        if(!filename)
+        if(!filename&&nextsize>3)
         {
           if((filename = xadAllocVec(XADM (xadUINT32) nextsize-3, XADMEMF_ANY)))
           {
@@ -140,7 +140,7 @@
         }
         break;
       case 0x02: /* Directoryname */
-        if(!dirname)
+        if(!dirname&&nextsize>3)
         {
           if((dirname = xadAllocVec(XADM (xadUINT32) nextsize-3, XADMEMF_ANY)))
           {
@@ -167,8 +167,7 @@
         }
         break;
       case 0x50: /* File permission */
-        xadConvertProtection(XADM XAD_PROTUNIX, EndGetI16(&bufptr[1]), XAD_GETPROTAMIGA,
-        &prot, TAG_DONE);
+        prot = EndGetI16(&bufptr[1]);
         break;
       case 0x51: /* ID's */
         groupid = EndGetI16(&bufptr[1]);
@@ -238,7 +237,10 @@
     TAG_IGNORE, commentsize, XAD_OBJPRIVINFOSIZE, sizeof(struct LhAPrivate)
     + groupsize + usersize, TAG_DONE)))
     {
-      fi->xfi_Protection = prot;
+      if(prot>=0)
+        xadConvertProtection(XADM XAD_PROTUNIX, prot, XAD_GETPROTFILEINFO,
+        &fi, TAG_DONE);
+
       if(time)
         xadConvertDates(XADM XAD_DATEUNIX, time, XAD_GETDATEXADDATE, &fi->xfi_Date, TAG_DONE);
       else
@@ -280,10 +282,15 @@
       }
       for(filename2 = fi->xfi_FileName; *filename2; ++filename2)
       {
+#ifdef NO_FILENAME_MANGLING
+        if((xadUINT8)*filename2 == 0xFF)
+          *filename2 = '/';
+#else
         if((xadUINT8)*filename2 == '\\' || (xadUINT8)*filename2 == 0xFF)
           *filename2 = '/';
         else if((*filename2&0x7F) <= 0x20 || *filename2 == 0x7F)
           *filename2 = '_';
+#endif
       }
       fi->xfi_Size = EndGetI32(&head[11]);
       fi->xfi_CrunchSize = EndGetI32(&head[7]) - (head[20] == 1 ? extsize : 0);
@@ -459,10 +466,15 @@
 
               for(i = 0; fi->xfi_FileName[i]; ++i)
               {
+#ifdef NO_FILENAME_MANGLING
+                if((xadUINT8)fi->xfi_FileName[i] == 0xFF)
+                  fi->xfi_FileName[i] = '/';
+#else
                 if((xadUINT8)fi->xfi_FileName[i] == 0xFF || (xadUINT8)fi->xfi_FileName[i] == '\\')
                   fi->xfi_FileName[i] = '/';
                 if((fi->xfi_FileName[i]&0x7F) < 0x20 || fi->xfi_FileName[i] == 0x7F)
                   fi->xfi_FileName[i] = '_';
+#endif
               }
 
               if(fi->xfi_Comment && (fi->xfi_Comment[0] == 'S' || fi->xfi_Comment[0] == 'U' ||
